smt.go

```diff
@@ -1,210 +1,392 @@
-    rootHash := smt.digest(rootValue)
-    ms.Put(rootHash, rootValue)
-    smt.SetRoot(rootHash)
-
-    return &smt
+	return &smt
 }
 
 // ImportSparseMerkleTree imports a Sparse Merkle tree from a non-empty MapStore.
 func ImportSparseMerkleTree(ms MapStore, hasher hash.Hash, root []byte) *SparseMerkleTree {
-    smt := SparseMerkleTree{
-        hasher: hasher,
-        ms: ms,
-        root: root,
-    }
-    return &smt
+	smt := SparseMerkleTree{
+		th:   *newTreeHasher(hasher),
+		ms:   ms,
+		root: root,
+	}
+	return &smt
 }
 
 // Root gets the root of the tree.
 func (smt *SparseMerkleTree) Root() []byte {
-    return smt.root
+	return smt.root
 }
 
 // SetRoot sets the root of the tree.
 func (smt *SparseMerkleTree) SetRoot(root []byte) {
-    smt.root = root
+	smt.root = root
 }
 
 func (smt *SparseMerkleTree) depth() int {
-    return smt.keySize() * 8
-}
-
-func (smt *SparseMerkleTree) keySize() int {
-    return smt.hasher.Size()
-}
-
-func (smt *SparseMerkleTree) defaultNode(height int) []byte {
-    return defaultNodes(smt.hasher)[height]
-}
-
-func (smt *SparseMerkleTree) digest(data []byte) []byte {
-    smt.hasher.Write(data)
-    sum := smt.hasher.Sum(nil)
-    smt.hasher.Reset()
-    return sum
+	return smt.th.pathSize() * 8
 }
 
 // Get gets a key from the tree.
 func (smt *SparseMerkleTree) Get(key []byte) ([]byte, error) {
-    value, err := smt.GetForRoot(key, smt.Root())
-    return value, err
+	value, err := smt.GetForRoot(key, smt.Root())
+	return value, err
 }
 
 // GetForRoot gets a key from the tree at a specific root.
 func (smt *SparseMerkleTree) GetForRoot(key []byte, root []byte) ([]byte, error) {
-    path := smt.digest(key)
-    currentHash := root
-    for i := 0; i < smt.depth(); i++ {
-        currentValue, err := smt.ms.Get(currentHash)
-        if err != nil {
-            return nil, err
-        }
-        if hasBit(path, i) == right {
-            currentHash = currentValue[smt.keySize():]
-        } else {
-            currentHash = currentValue[:smt.keySize()]
-        }
-    }
-
-    value, err := smt.ms.Get(currentHash)
-    if err != nil {
-        return nil, err
-    }
-
-    return value, nil
-}
-
-// Update sets a new value for a key in the tree, returns the new root, and sets the new current root of the tree.
+	if bytes.Equal(root, smt.th.placeholder()) {
+		// The tree is empty, return the default value.
+		return defaultValue, nil
+	}
+
+	path := smt.th.path(key)
+	currentHash := root
+	for i := 0; i < smt.depth(); i++ {
+		currentData, err := smt.ms.Get(currentHash)
+		if err != nil {
+			return nil, err
+		} else if smt.th.isLeaf(currentData) {
+			// We've reached the end. Is this the actual leaf?
+			p, valueHash := smt.th.parseLeaf(currentData)
+			if !bytes.Equal(path, p) {
+				// Nope. Therefore the key is actually empty.
+				return defaultValue, nil
+			}
+			// Otherwise, yes. Return the value.
+			value, err := smt.ms.Get(valueHash)
+			if err != nil {
+				return nil, err
```

Comment:
These `Get`s won't err in tests because the in-memory store never errs. In the future we might want to test these too.

Reply:
Sure, but if someone changes the in-memory store implementation in the future then this ensures that any errors returned are caught in tests.